packages/c/sshnpd/src/device_info.c

```diff
@@ -0,0 +1,166 @@
+#include "atlogger/atlogger.h"
+#include "sshnpd/daemon.h"
+#include "sshnpd/params.h"
+#include <atclient/atclient.h>
+#include <sshnpd/device_info.h>
+#include <string.h>
+#include <time.h>
+
+int handle_username_keys(atclient *atclient, const char **atsigns, size_t num_atsigns, const char *username,
+                         const char *device_name, const char *device_atsign, bool make_visible) {
+  int ret;
+  const char *TAG = "send_username_keys";
+
+  if (num_atsigns <= 0) {
+    return 0;
+  }
+
+  for (size_t i = 0; i < num_atsigns; i++) {
+    const char *atsign = atsigns[i];
+    // example: @client:username.devicename.sshnp@device
+    size_t atkey_strlen = strlen(atsign) + strlen(device_name) + strlen(device_atsign) + 17;
+    char atkey_str[atkey_strlen];
+    snprintf(atkey_str, atkey_strlen, "%s:username.%s.sshnp%s", atsign, device_name, device_atsign);
+
+    atclient_atkey atkey;
+    atclient_atkey_init(&atkey);
+    ret = atclient_atkey_from_string(&atkey, atkey_str);
+    if (ret != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to build username key for %s\n", atsign);
+      continue;
+    }
+
+    atclient_atkey_metadata *metadata = &atkey.metadata;
+    ret = atclient_atkey_metadata_set_is_public(metadata, false);
+    if (ret != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set is_public on metadata for %s username key\n",
+                   atsign);
+      atclient_atkey_free(&atkey);
+      continue;
+    }
+
+    ret = atclient_atkey_metadata_set_is_encrypted(metadata, true);
+    if (ret != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set is_encrypted on metadata for %s username key\n",
+                   atsign);
+      atclient_atkey_free(&atkey);
+      continue;
+    }
+    ret = atclient_atkey_metadata_set_ttr(metadata, -1);
+    if (ret != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set ttr on metadata for %s username key\n", atsign);
+      atclient_atkey_free(&atkey);
+      continue;
+    }
+    ret = atclient_atkey_metadata_set_ccd(metadata, true);
+    if (ret != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set ccd on metadata for %s username key\n", atsign);
+      atclient_atkey_free(&atkey);
+      continue;
+    }
+    if (make_visible) {
+      ret = atclient_put_shared_key(atclient, &atkey, username, NULL, NULL);
+      atclient_atkey_free(&atkey);
+      if (ret != 0) {
+        atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to put username key for %s\n", atsign);
+        continue;
+      }
+    } else {
+      ret = atclient_delete(atclient, &atkey, NULL, NULL);
+      atclient_atkey_free(&atkey);
+      if (ret != 0) {
+        atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to delete username key for %s\n", atsign);
+        continue;
+      }
+    }
+  }
+
+  return ret;
+}
+
+void send_next_device_info(atclient *atclient, sshnpd_params *params) {
```

Comment:
Function which puts the next device info key if an hour has passed since the last time it was put by the current process (not determined by atkey metadata).

Reply:
The downside to beware of is that each iteration of this is between monitor timeouts (5 seconds), so if there are a ton of manager atsigns, it will take `5s * number of managers` to fully update on startup.